executor/include/execution/trait/is_executor.hpp

```diff
@@ -4,24 +4,31 @@
 
 #pragma once
 
+#include <execution/executor/base_executor.hpp>
 #include <experimental/type_traits>
 #include <type_traits>
 
 namespace execution {
 
-template <class Executor, typename = void>
+template <class Executor, typename Function = void (*)(), typename = void>
 struct is_executor : std::false_type {};
 
-constexpr const auto noop = [] {};
+namespace detail {
+const auto noop = [] {};
+}
 
-template <class Executor>
+template <typename Executor, typename Function>
 struct is_executor<
-    Executor,
+    Executor, Function,
     std::enable_if_t<
-        std::is_same_v<void, decltype(std::declval<Executor>().execute(noop))>,
-        void>> : std::true_type {};
+        std::is_copy_constructible_v<
+            std::remove_cv_t<std::remove_reference_t<Executor>>> &&
+            std::is_same_v<
+                decltype(std::declval<Executor>().execute(detail::noop)), void>,
+        decltype(std::declval<Executor&>() == std::declval<Executor&>(),
+                 (void)0)>> : std::true_type {};
```

Comment:
Why is `void(0)` needed?

Reply:
Why is `(void)0` needed?

Reply:
To return the void type if the expression `std::declval<Executor&>() == std::declval<Executor&>()` is valid.
This article explains it better https://stackoverflow.com/questions/16044514/what-is-decltype-with-two-arguments

Just syntactic sugar in the end.

Reply:
Please rethink the way you've constructed the template. My concerns are:
* `is_executor` accepts a `Function` but doesn't test it. Is the function even needed? `is_executor_v` doesn't seem to need or expose the function
* The condition needs to be broken in 2 parts of `enable_if`
* Conditions are hard to understand. They should be refactored into named requirements so it's easier to read and understand

```cpp

template <class Candidate>
struct is_executor: std::conditional<
detail::all_objects_are_equal<Candidate> && detail::accepts_void_return_no_param_functor<Candidate>
,std::true_type, std::false_type> {};
```

My names are ofc bad (the papers you've read might have better names), but I hope this shows some direction which would be easier to understand